File: algorithms/11_golang/sorting/counting_sort.go
package main

import "fmt"

func countingSort(a []int, maxVal int) []int {
	count := make([]int, maxVal+1)
	for _, v := range a {
		count[v]++
	}
	for i := 1; i <= maxVal; i++ {
		count[i] += count[i-1]
	}
	out := make([]int, len(a))
	for i := len(a) - 1; i >= 0; i-- {
		v := a[i]
		count[v]--
		out[count[v]] = v
	}
	return out
}

func main() {
	data := []int{4, 2, 2, 8, 3, 3, 1}
	sorted := countingSort(data, 8)
	want := []int{1, 2, 2, 3, 3, 4, 8}
	for i := range want {
		if sorted[i] != want[i] {
			panic("counting sort: mismatch")
		}
	}
	fmt.Println("[Go CountingSort] Counting sort verified:", sorted)
}

File: algorithms/21_scala/sorting/CountingSort.scala
object CountingSort {
  def countingSort(arr: Array[Int]): Array[Int] = {
    if (arr.isEmpty) return Array.emptyIntArray
    val maxVal = arr.max
    val count = Array.fill(maxVal + 1)(0)
    arr.foreach(x => count(x) += 1)
    for (i <- 1 to maxVal) count(i) += count(i - 1)
    val out = new Array[Int](arr.length)
    arr.reverse.foreach { x =>
      count(x) -= 1
      out(count(x)) = x
    }
    out
  }

  def main(args: Array[String]): Unit = {
    val data = Array(4, 2, 2, 8, 3, 3, 1)
    val sorted = countingSort(data)
    assert(sorted.sameElements(data.sorted), "not sorted")
    println(s"[Scala CountingSort] Counting sort verified: ${sorted.mkString(", ")}")
  }
}
